cpu/esp_common: Port to inline-able IRQ API

maribu opened the pull request:
### Contribution description

Port ESP32/ESP8266 to the new `inline`-able IRQ API

### Testing procedure

#### `tests/irq_disable_restore`

I ran this on the `esp32-mh-et-live-minikit`. Someone should at least test with an ESP8266 as well (I have no ESP8266 at hand right now).

#### `tests_bench_thread_yield_pingpong`

ROM: 32 bytes less (some more due to longer branch name in message)
Speed: ~4.6% faster

<details><summary>With <code>master</code></summary>

```
   text	   data	    bss	    dec	    hex	filename
  56762	   3648	   9520	  69930	  1112a	/data/riotbuild/riotbase/tests/bench_thread_yield_pingpong/bin/esp32-mh-et-live-minikit/tests_bench_thread_yield_pingpong.elf

2020-06-26 11:55:36,241 # Help: Press s to start test, r to print it is ready
s
2020-06-26 11:55:57,764 # START
2020-06-26 11:55:57,767 # main(): This is RIOT! (Version: 2020.07-devel-1580-g279f2)
2020-06-26 11:55:57,768 # main starting
2020-06-26 11:55:58,760 # { "result" : 50505 }
```

</details>

<details><summary>This PR</summary>

```   text	   data	    bss	    dec	    hex	filename
  56730	   3648	   9520	  69898	  1110a	/data/riotbuild/riotbase/tests/bench_thread_yield_pingpong/bin/esp32-mh-et-live-minikit/tests_bench_thread_yield_pingpong.elf

2020-06-26 11:52:20,507 # Help: Press s to start test, r to print it is ready
s
2020-06-26 11:52:23,046 # START
2020-06-26 11:52:23,050 # main(): This is RIOT! (Version: 2020.07-devel-1581-g375faa-esp-inline-irq)
2020-06-26 11:52:23,051 # main starting
2020-06-26 11:52:24,041 # { "result" : 52840 }

```

</details>

#### `tests_bench_mutex_pingpong`

ROM: 36 bytes less (some more due to longer branch name)
Speed: ~5.0% faster

<details><summary>With <code>master</code></summary>

```  text	   data	    bss	    dec	    hex	filename
  56758	   3648	   9528	  69934	  1112e	/data/riotbuild/riotbase/tests/bench_mutex_pingpong/bin/esp32-mh-et-live-minikit/tests_bench_mutex_pingpong.elf

2020-06-26 11:57:29,511 # Help: Press s to start test, r to print it is ready
s
2020-06-26 11:57:42,916 # START
2020-06-26 11:57:42,917 # main(): This is RIOT! (Version: 2020.07-devel-1580-g279f2)
2020-06-26 11:57:42,917 # main starting
2020-06-26 11:57:43,912 # { "result" : 44692 }

```

</details>

<details><summary>This PR</summary>

```
   text	   data	    bss	    dec	    hex	filename
  56722	   3648	   9528	  69898	  1110a	/data/riotbuild/riotbase/tests/bench_mutex_pingpong/bin/esp32-mh-et-live-minikit/tests_bench_mutex_pingpong.elf

2020-06-26 11:58:24,901 # Help: Press s to start test, r to print it is ready
s
2020-06-26 11:58:35,389 # START
2020-06-26 11:58:35,393 # main(): This is RIOT! (Version: 2020.07-devel-1581-g375faa-esp-inline-irq)
2020-06-26 11:58:35,394 # main starting
2020-06-26 11:58:36,384 # { "result" : 46948 }
```

</details>

### Issues/PRs references

Tracking issue: https://github.com/RIOT-OS/RIOT/issues/14356

gschorcht commented:
@maribu I already had a local version of these changes here. I had used them to determine memory sizes, which I published yesterday in issue #14356. I didn't open a PR since it wasn't worth from my point of view. So we have done it twice :wink: Anyway.

Some first comments:

1. You have now an additional header file `cpu/esp_common/irq_arch.h` which defines all `irq_*` functions again but not inlined. For which purpose?
2. The implementation of `irq_restore` is tricky and I'm not really comfortable with it since it is an switch on/off solution. The Xtensa architecture for ESP32 define 32 interrupts using 7 different levels. The details are shown in the table below.

   Even though RIOT itself doesn't use different interrupt levels, we use binary libraries from the SDK that might rely on interrupts of different interrupt levels. That's why the original version of `irq_disable` doesn't disable all interrupt levels, but only those that are lower than or equal to `XCHAL_EXCM_LEVEL`. `XCHAL_EXCM_LEVEL` for ESP32 is 3, that is interrupts at level 4 to 15 including the NMI keep active. NMI is used for the WiFi interface for example. So simply disabling all interrupts might lead to stability problems with WiFi. Furthermore, we know that interrupt 25 (level 4) is used for cache errors and interrupt 28 (level 4) is used for DPORT access. Disabling these interrupts might also affect the stability.

   Furthermore, what if `irq_disable` is called when the interrupt level is set to 2 by any binary library and `irq_restore` sets it simply to 15?
3. Compilation of a ESP8266 leads to a lot of multiple defined error.

| No | Category | Type | Priority Level |
|:----|:----|:----|:----- |
| 0 | Peripheral | Level-Triggered | 1 |
| 1 | Peripheral | Level-Triggered | 1 |
| 2 | Peripheral | Level-Triggered | 1 |
| 3 | Peripheral | Level-Triggered | 1 |
| 4 | Peripheral | Level-Triggered | 1 |
| 5 | Peripheral | Level-Triggered | 1 |
| 6 | Internal | Timer.0 | 1 |
| 7 | Internal | Software | 1 |
| 8 | Peripheral | Level-Triggered | 1 |
| 9 | Peripheral | Level-Triggered | 1 |
| 10 | Peripheral | Edge-Triggered | 1 |
| 11 | Internal | Profiling | 3 |
| 12 | Peripheral | Level-Triggered | 1 |
| 13 | Peripheral | Level-Triggered | 1 |
| 14 | Peripheral | NMI | NMI |
| 15 | Internal | Timer.1 | 3 |
| 16 | Internal | Timer.2 | 5 |
| 17 | Peripheral | Level-Triggered | 1 |
| 18 | Peripheral | Level-Triggered | 1 |
| 19 | Peripheral | Level-Triggered | 2 |
| 20 | Peripheral | Level-Triggered | 2 |
| 21 | Peripheral | Level-Triggered | 2 |
| 22 | Peripheral | Edge-Triggered | 3 |
| 23 | Peripheral | Level-Triggered | 3 |
| 24 | Peripheral | Level-Triggered | 4 |
| 25 | Peripheral Level-Triggered | 4 |
| 26 | Peripheral | Level-Triggered | 5 |
| 27 | Peripheral | Level-Triggered | 3 |
| 28 | Peripheral | Edge-Triggered | 4 |
| 29 | Internal | Software | 3 |
| 30 | Peripheral | Edge-Triggered | 4 |
| 31 | Peripheral | Level-Triggered | 5 |

gschorcht commented on cpu/esp_common/include/irq_arch_common.h:
```diff
@@ -67,6 +67,79 @@ extern volatile uint32_t irq_interrupt_nesting;
+        /* this will block IRQs with prio <= intlevel */
+        "rsil %[state], %[intlevel]"
+        : [state]       "=a" (state)
+        : [intlevel]    "I"(15) /* 15 is highest prio, blocking all IRQs */
```
```suggestion
        : [intlevel]    "I"(XCHAL_EXCM_LEVEL) /* 15 is highest prio, blocking all IRQs */
```

gschorcht replied:
We should use a constant from the SOC definition here. We could use `XCHAL_NUM_INTLEVELS` to disable all interrupts or `XCHAL_EXCM_LEVEL` to disable only low level interrupts.

ESP32:
```
#define XCHAL_EXCM_LEVEL		3
#define XCHAL_NUM_INTLEVELS		6
```
ESP8266:
```
#define XCHAL_EXCM_LEVEL		1
#define XCHAL_NUM_INTLEVELS		2
```

gschorcht commented on cpu/esp_common/include/irq_arch_common.h:
```diff
@@ -67,6 +67,79 @@ extern volatile uint32_t irq_interrupt_nesting;
+     * or disables *all* IRQs. So if state is 0 (all IRQs enabled), we will
+     * do so using rsil, otherwise we don't touch the interrupt level
+     */
+    if (!state) {
```
With this implementation, you can't restore interrupt levels that are greater than 0. This situation could occur if `irq_disable` is called while the `PS.INTLEVEL` register is already greater than 0 for some reason, e.g., because the low level interrupts have already been disabled by SDK code that we do not control.

Example: `periph/timer` uses interrupt 19 (level 2). Let's suppose that the SDK code has only disabled low level interrupts (level 1) what is really possible. Then the `PS.INTLEVEL` register is 1. If the timer then triggers an interrupt, RIOT code is executed. That is, `irq_disable` is called on entry with `PS.INTLEVEL=1` and `irq_restore` is called on exit. However, `irq_disable` wouldn't restore the old interrupt level and all interrupts would remain disabled.

I don't know how probable it is, but it can happen in theory. If it did happen, we would sporadically get errors that we all "love" so much because they are not reproducible. As this change could have a major impact on stability, I would suggest not making this change. I would sacrifice performance in favor of stability.

maribu commented:
> 1. You have now an additional header file `cpu/esp_common/irq_arch.h` which defines all `irq_*` functions again but not inlined. For which purpose?

A leftover from development that sneaked in :-/ I dropped it now.

> 2. The implementation of `irq_restore` is tricky and I'm not really comfortable with it since it is an switch on/off solution.

But that is *exactly* what the RIOTs IRQ API should do. At the very least all IRQs should be blocked whose ISR might call into RIOT code, as otherwise data corruptions are bound to occur.

>  NMI is used for the WiFi interface for example. So simply disabling all interrupts might lead to stability problems with WiFi.

This is not true. RIOT code is developed with the expectation that all IRQs are disabled between `irq_disable()` ...`irq_restore()`. Thus, every RIOT developer takes great care to keep those critical sections as short as possible. If there were long critical sections in RIOT protected by `irq_disable()` ... `irq_restore()`, those would have lead to RIOT becoming unusable for real time applications. But e.g. the MAC guys will insist that RIOT remains able to meet timing constraints of time slotted protocols. Wifi should not be much different in that regard.

> Furthermore, we know that interrupt 25 (level 4) is used for cache errors

So ROM access is managed in software, not hardware? Doesn't this have a major impact on its suitability for real time applications, as the CPU could be busy loading the next ROM page into cache for a low prio task and be unable to perform a high prio IRQ while doing so?

But you're right, we certainly cannot assume that every `irq_disable()` ... `irq_restore()` sequence will always be on the same ROM page and must never block such IRQs from RIOT.

gschorcht commented:
> > 2. The implementation of `irq_restore` is tricky and I'm not really comfortable with it since it is an switch on/off solution.
> 
> But that is _exactly_ what the RIOTs IRQ API should do. At the very least all IRQs should be blocked whose ISR might call into RIOT code, as otherwise data corruptions are bound to occur.
 
I know. Unfortunately, the RIOT code must coexist and work together with the binary SDK code, which assumes that it has interrupt levels. Please refer the example in comment https://github.com/RIOT-OS/RIOT/pull/14370#discussion_r446195354. There, I have constructed a possible case where `irq_restore` wouldn't work.

> > NMI is used for the WiFi interface for example. So simply disabling all interrupts might lead to stability problems with WiFi.
> 
> This is not true. RIOT code is developed with the expectation that all IRQs are disabled between `irq_disable()` ...`irq_restore()`. Thus, every RIOT developer takes great care to keep those critical sections as short as possible. If there were long critical sections in RIOT protected by `irq_disable()` ... `irq_restore()`, those would have lead to RIOT becoming unusable for real time applications. But e.g. the MAC guys will insist that RIOT remains able to meet timing constraints of time slotted protocols. Wifi should not be much different in that regard.
> 
> > Furthermore, we know that interrupt 25 (level 4) is used for cache errors
> 
> So ROM access is managed in software, not hardware?

The cache is managed in hardware. Cache error interrupt raises if cached ROM is tried to be accessed but the cache is not enabled. Found here https://github.com/espressif/esp-idf/blob/12b497b328b95f83607e378b7e14c968ed3d1183/components/esp32/cache_err_int.c#L15-L20

Finally, my biggest concern is that the sequence `irq_disable()` ... `irq_restore()` could be called while the SDK has only disabled low-level interrupts (level 1). After the return of `irq_restore()` all interrupts would be disabled.

gschorcht commented:
> #### `tests_bench_thread_yield_pingpong`
> ROM: 32 bytes less (some more due to longer branch name in message)

You might compile with `RIOT_VERSION="" CFLAGS='-DNDEBUG' ` to have more accurate sizes.

gschorcht commented:
It also works with ESP8266. See the benchmarks below.

#### tests/bench_thread_yield_pingpong
ROM: 160 bytes less
Speed: almost 10 % faster
<details><summary>With Master</summary>

```
   text	   data	    bss	    dec	    hex	filename
 278128	   5068	  31808	 315004

{ "result" : 62499 }
```
</details>

<details><summary>This PR</summary>

```
   text	   data	    bss	    dec	    hex	filename
 277948	   5068	  31808	 314824

{ "result" : 68728 }
```
</details>

#### tests/bench_mutex_pingpong
ROM: 160 bytes less
Speed: 9,1 % faster
<details><summary>With Master</summary>

```
   text	   data	    bss	    dec	    hex	filename
 278112	   5068	  31816	 314996```

{ "result" : 55325 }
```
</details>

<details><summary>This PR</summary>

```
   text	   data	    bss	    dec	    hex	filename
 277948	   5068	  31816	 314832```

{ "result" : 60606 }
```
</details>

gschorcht commented:
I checked again the interrupts that are really used. Although interrupt 25 (level 4) is defined for cache errors by the SDK, it is not used in RIOT. Interrupt 28 (level 4) that is defined for DPORT access, would only be important if both cores are used which is not the case in RIOT. That is, these high-level interrupts are not a problem if they are disabled.

gschorcht commented:
I could find a reproducible situation where `irq_disable` is called with `PS.INTLEVEL=1`. This happens whenever `irq_disable` is called within an ISR, which is probably often the case. The reason for this is that higher level interrupts are enabled in the interrupt level 1 handler before the ISR is called. https://github.com/RIOT-OS/RIOT/blob/1d6a83c79746e749ce76e9aefa05c0e08f8ff3e4/cpu/esp_common/vendor/xtensa/xtensa_vectors.S#L1307-L1317 Calling `irq_disable` and `irq_enable` sequence inside an ISR, would leave interrupts at all levels disabled with this PR.
Fortunately, when entering the interrupt level 1 handler, the PS register is first stored on the stack https://github.com/RIOT-OS/RIOT/blob/1d6a83c79746e749ce76e9aefa05c0e08f8ff3e4/cpu/esp_common/vendor/xtensa/xtensa_vectors.S#L1288-L1293 and is restored on exit https://github.com/RIOT-OS/RIOT/blob/1d6a83c79746e749ce76e9aefa05c0e08f8ff3e4/cpu/esp_common/vendor/xtensa/xtensa_vectors.S#L981-L983

The question is whether we can be sure that there is no situation where the interrupts are locked by `irq_enabled` even though they should be enabled. The situation described above is perhaps only one that I have found and there are others.

maribu commented:
I wonder if it would make sense to have `irq_enable()` / `irq_disable()` assuming that IRQs are indeed binary (on / off), but instead have in the ISR trampolines a multi-level `irq_disable()` / `irq_restore()`.

Specifically, whenever an ISR is called it will disable IRQs to the highest level that can call into RIOT code and the previous level is backed up. Once the RIOT IRQ callback function returns to the ISR, the ISR will restore the previous IRQ state.

I think this would result in RIOT code always being executed with either no IRQs blocked, or all IRQs relevant to RIOT being blocked. So that way `irq_restore()` and `irq_disable()` can indeed be binary.

gschorcht commented:
> Specifically, whenever an ISR is called it will disable IRQs to the highest level that can call into RIOT code and the previous level is backed up. Once the RIOT IRQ callback function returns to the ISR, the ISR will restore the previous IRQ state.

Do you mean that the callback in `_uart_intr_handler` https://github.com/RIOT-OS/RIOT/blob/06dca07b0a03dd034f85719a1cf7a5266d0aac86/cpu/esp_common/periph/uart.c#L263 should look like the following, for example
```diff
+    uint32_t level = irq_int_level_save();
     _uarts[uart].isr_ctx.rx_cb(_uarts[uart].isr_ctx.arg, data);
+    irq_int_level_restore(level);
```
where `irq_int_level_save` and `irq_int_level_restore` would be the former implementations of `irq_disable` and `irq_enable`?

gschorcht commented:
OK, ISRs are just one case to consider. The other cases are RIOT functions which are called back from SDK libraries, for example when a packet is received over the WiFi interface. At least for the WiFi interface, it is known that the RIOT function is called back in thread context with interrupts enabled.

For other callbacks, the fact might help that ESP-IDF is based on FreeRTOS and therefore expects that interrupts are either all disabled with `portDISABLE_INTERRUPTS` or all enabled with `portENABLE_INTERRUPTS`.

On the other hand, the ESP-IDF port of FreeRTOS defines and uses the additional macros `portENTER_CRITICAL_NESTED` and `portEXIT_CRITICAL_NESTED` which do exactly what our `irq_disable` and `irq_restore` functions did. That's why these macros were simply mapped to `irq_disable` and `irq_restore` in RIOT.

According to the comments in [components/freertos/include/freertos/portmacro.h](https://github.com/espressif/esp-idf/blob/release/v3.1/components/freertos/include/freertos/portmacro.h#L238) `portDISABLE_INTERRUPTS` and `portENABLE_INTERRUPTS`  should be used with a lot of care and cannot be called from interrupt level :thinking: `portENTER_CRITICAL_NESTED` and `portEXIT_CRITICAL_NESTED` are required for nested interrupts. I don't know how often nested interrupts might happen.

IMHO, in current ESP32 RIOT implementation, a nested interrupt could only happen within an level 1 ISR while the timer interrupt of level 2 occurs.

So we have two options
1. We use the binary (on/off) IRQ disable/enable as it is without worrying about anything else. In this case, the PR would be ok as it is. BTW, with your implementation, I have been running a WiFi connection on an ESP32 for almost a day.
2. To be absolutely sure, we could save/restore the interrupt level on every entry/exit to/from ISRs and all callbacks. However, this would be a bigger change. I suppose we would have to do it for all functions that are realized in RIOT and used [cpu/esp32/vendor/esp-idf/esp32/wifi_os_adapter.c](https://github.com/RIOT-OS/RIOT/blob/master/cpu/esp32/vendor/esp-idf/esp32/wifi_os_adapter.c) and [cpu/esp8266/vendor/esp-idf/esp8266/source/esp_wifi_os_adapter.c](https://github.com/RIOT-OS/RIOT/blob/master/cpu/esp8266/vendor/esp-idf/esp8266/source/esp_wifi_os_adapter.c)

BTW, `portDISABLE_INTERRUPT` uses `XCHAL_EXCM_LEVEL` to disable interrupts.

gschorcht commented:
@maribu Do we want to reopen this PR? I guess it is still an action item.

maribu commented:
I haven't come up with an implementation that addresses all issues and is worth having inlined. For all other platforms disabling and restoring IRQs wasso trivial that less code was required in the implementation compared to the function call.

Maybe it is indeed just a better trade-off to keep the implementation as proper C function. (And I do hope that LTO will be mandatorily enabled for all platforms and toolchains in the not to distant future, so that we don't need to have `static inline` functions at all.)
